### DISCUSSION TITLE
MariaDB Won't Start Up On Boot (Catalina)

### DISCUSSION THREAD

**Discussion by l008com** [Homebrew/discussions#3074] [Getting started] [answered]
I have a nice 2012 Mac mini running Catalina. I used homebrew to install MariaDB. The DB works great once it's running, but I have been unable to get it to launch on boot. 

I tried using brew services, which makes a launchd to start up the database on boot. It doesn't work. 
I made my own launchd.plist to try to launch it on boot. I'm very familiar with these launchd scripts. Mine runs the same command as the homebrew script, and importantly, the same exact command I use to *successfully* launch MariaDB from the terminal, once the Mac is booted. But the launchd just won't work. 

I set a stdout and stderr for the launchd, there are no errors in them. I look at the mariadb error log, theres nothing there. It just quietly fails every single time. 

As you can imagine, this is a huge problem because any time the server reboots, I have to manually log on and start the database, else the web things this server hosts will not work properly. 

Here is the full launchd.plist I use, minus the stdout/stderr stuff:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-AppleDTD PLIST 1.0EN" "[http://www.apple.com/DTDs/PropertyList-1.0.dtd"](http://www.apple.com/DTDs/PropertyList-1.0.dtd%22)>
<plist version="1.0">
<dict>
<key>Label</key>
<string>home-server.mariadb</string>
<key>ProgramArguments</key>
<array>
<string>/usr/local/opt/mariadb/bin/mysqld_safe</string>
<string>--datadir=/usr/local/var/mysql</string>
</array>
<key>RunAtLoad</key>
<true/>
<key>WorkingDirectory</key>
<string>/usr/local/var</string>
</dict>
</plist>
```

That script is simply running `sudo /usr/local/opt/mariadb/bin/mysqld_safe --datadir=/usr/local/var/mysql` as system startup. Yet it doesn't work. Note that the launchd is running it as root because the script is in /Library/LaunchDeamons, not ~/Library...

I was righting with this issue about 8 months ago when I set up this machine. I never find any solutions or even any solid clues and gave up. But it's really a pain in the ass so I'd really like to find a solution to the problem.

**Comment by l008com** [comment]
I had a seemingly identical problem a few years back on an El Capitan machine. That time, the problem turned out to be MariaDB failing because it's data directly was owned by the wrong user. And that caused it to fail silently. That's not the case here though, the user is correct and the server will run..... just not automatically at boot.

**Comment by carlocab** [comment]
> I tried using brew services, which makes a launchd to start up the database on boot. It doesn't work.

This really isn't a lot to go on to try to understand the issue you're having. Could you elaborate what steps you tried to do this, and what "it doesn't work" means here, exactly?

**Comment by l008com** [comment]
The following is what I get in the MariaDB error log when I launch it manually from the terminal:

`210721 23:58:32 mysqld_safe Starting mariadbd daemon with databases from /usr/local/var/mysql`
`2021-07-21 23:58:33 0 [Note] /usr/local/opt/mariadb/bin/mariadbd (mysqld 10.6.3-MariaDB) starting as process 1124 ...`
`2021-07-21 23:58:33 0 [Warning] Setting lower_case_table_names=2 because file system for /usr/local/var/mysql/ is case insensitive`
`2021-07-21 23:58:33 0 [Note] InnoDB: Compressed tables use zlib 1.2.11`
`2021-07-21 23:58:33 0 [Note] InnoDB: Number of pools: 1`
`2021-07-21 23:58:33 0 [Note] InnoDB: Using crc32 + pclmulqdq instructions`
`2021-07-21 23:58:33 0 [Note] InnoDB: Initializing buffer pool, total size = 134217728, chunk size = 134217728`
`2021-07-21 23:58:33 0 [Note] InnoDB: Completed initialization of buffer pool`
`2021-07-21 23:58:33 0 [Note] InnoDB: 128 rollback segments are active.`
`2021-07-21 23:58:33 0 [Note] InnoDB: Creating shared tablespace for temporary tables`
`2021-07-21 23:58:33 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ...`
`2021-07-21 23:58:33 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB.`
`2021-07-21 23:58:33 0 [Note] InnoDB: 10.6.3 started; log sequence number 17013088; transaction id 146078`
`2021-07-21 23:58:33 0 [Note] Plugin 'FEEDBACK' is disabled.`
`2021-07-21 23:58:34 0 [Note] InnoDB: Loading buffer pool(s) from /usr/local/var/mysql/ib_buffer_pool`
`2021-07-21 23:58:34 0 [Note] Server socket created on IP: '::'.`
`2021-07-21 23:58:34 0 [Note] Server socket created on IP: '0.0.0.0'.`
`2021-07-21 23:58:34 0 [Note] InnoDB: Buffer pool(s) load completed at 210721 23:58:34`
`2021-07-21 23:58:34 0 [ERROR] Incorrect definition of table mysql.event: expected column 'definer' at position 3 to have type varchar(, found type char(141).`
`2021-07-21 23:58:34 0 [ERROR] mariadbd: Event Scheduler: An error occurred when initializing system tables. Disabling the Event Scheduler.`
`2021-07-21 23:58:34 0 [Note] /usr/local/opt/mariadb/bin/mariadbd: ready for connections.`
`Version: '10.6.3-MariaDB'  socket: '/tmp/mysql.sock'  port: 3306  Homebrew`

And this is what I get in the MariaDB error log when launchd tries to start MariaDB automatically as system boot:

`210721 23:51:49 mysqld_safe Starting mariadbd daemon with databases from /usr/local/var/mysql`
`2021-07-21 23:51:49 0 [Note] /usr/local/opt/mariadb/bin/mariadbd (mysqld 10.6.3-MariaDB) starting as process 462 ...`
`2021-07-21 23:51:49 0 [Warning] Setting lower_case_table_names=2 because file system for /usr/local/var/mysql/ is case insensitive`
`/usr/local/opt/mariadb/bin/mariadbd: Please consult the Knowledge Base to find out how to run mysqld as root!`
`2021-07-21 23:51:49 0 [ERROR] Aborting`
`210721 23:51:49 mysqld_safe mysqld from pid file /usr/local/var/mysql/Home-Server.local.pid ended`

That is the only feedback I've been able to find. 
Note that I'm not trying to run MariaDB as root. I'm trying to run it as _mysql, which is how it runs when I start it from the terminal, using the exact same command.

**Comment by l008com** [comment, marked as answer]
I found a solution! I'm not sure how or why it works, but if you add 'sudo' to the command launchd runs, mariadb is able to start up normally! This doesn't make sense because these launchdeamons are already being run as root, so adding sudo shouldn't DO anything. Yet, it fixes it! Go figure.

```
...
<key>ProgramArguments</key>
	<array>
		<string>/usr/bin/sudo</string>
		<string>/usr/local/opt/mariadb/bin/mariadbd-safe</string>
		<string>--datadir=/usr/local/var/mysql</string>
	</array>
...
```
